onnxruntime/core/optimizer/conv_add_act_fusion.cc

```diff
@@ -0,0 +1,287 @@
+    if (graph_utils::IsSupportedOptypeVersionAndDomain(*add_node, "Add", {7, 13, 14})) {
+      conv_node = SelectProducerConv(*add_node);
+    }
+    if (!conv_node) {
+      return std::nullopt;
+    }
+    const auto* act_node = GetLoneConsumerNode(graph_viewer, *add_node);
+    // even the next node is not a activation node, it's also fine.
+    if (!act_node || add_node->GetOutputEdgesCount() > 1) {
+      // we can't fuse add-activation when add_node has multiple consumer nodes
+      act_node = nullptr;
+    } else if (SelectActivation(graph_viewer, *act_node)) {
+      // this branch is deliberately empty as we want to keep 'act_node' as remains.
+    } else {
+      act_node = nullptr;
+    }
+
+    NodesToOptimizeIndicesBuilder builder{};
+    builder.target_node = conv_node->Index();
+    builder.output_nodes = {add_node->Index()};
+    if (act_node) {
+      builder.output_nodes.push_back(act_node->Index());
+    }
+    return builder.Build();
+  }
+
+  static bool SelectActivation(const GraphViewer& graph_viewer, const Node& activation_node) {
+    auto is_supported_cpu_ep_activation = [&graph_viewer](const Node& activation_node) {
+      if (graph_utils::IsSupportedOptypeVersionAndDomain(activation_node, "Relu", {6, 13, 14}) ||
+          graph_utils::IsSupportedOptypeVersionAndDomain(activation_node, "Sigmoid", {6, 13}) ||
+          graph_utils::IsSupportedOptypeVersionAndDomain(activation_node, "Tanh", {6, 13}) ||
+          graph_utils::IsSupportedOptypeVersionAndDomain(activation_node, "LeakyRelu", {6})) {
+        return true;
+      }
+
+      if (graph_utils::IsSupportedOptypeVersionAndDomain(activation_node, "Clip", {6, 11, 12, 13})) {
+        float min, max;
+        if (!optimizer_utils::GetClipConstantMinMax(graph_viewer.GetGraph(), activation_node, min, max)) {
+          return false;
+        }
+        return true;
+      }
+
+      if (graph_utils::IsSupportedOptypeVersionAndDomain(activation_node, "HardSigmoid", {6})) {
+        return true;
+      }
+      return false;
+    };
+    return is_supported_cpu_ep_activation(activation_node);
+  }
+
+  const Node* SelectProducerConv(const Node& node) const {
+    InlinedVector<const Node*> inputs_node;
+    constexpr int32_t kTensorDims = 4;  // NCHW
+    const auto& input_defs = node.InputDefs();
+
+    for (auto producer_node_ptr = node.InputNodesBegin(); producer_node_ptr != node.InputNodesEnd(); ++producer_node_ptr) {
+      const Node* producer_node = dynamic_cast<const Node*>(&(*producer_node_ptr));
+      inputs_node.push_back(producer_node);
+    }
+    size_t input_defs_count = input_defs.size();
+    // Test if all of inputs have an equal shape.
+    bool all_shapes_match = true;
+    auto* input_0_shape = input_defs[0]->Shape();
+    for (size_t n = 1; n < input_defs_count; n++) {
+      // Check if ONNX shape inferencing has computed a precise dimension value.
+      if (input_0_shape->dim_size() != kTensorDims) {
+        all_shapes_match = false;
```

Comment:
>all_shapes_match = false; [](http://example.com/codeflow?start=8&length=25)

return nullptr directly for all the set of false

Reply:
It should be fine with a final-return